### PR TITLE
[JsonPatch]: Move exception messages to resources

### PR DESCRIPTION
**Pull request by kirthik**
Fix for https://github.com/aspnet/Mvc/issues/2215 

@rynowak - Could you please take a look?

### CODE REVIEW

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 1]
```diff
@@ -158,20 +158,18 @@ private void Add(string path, object value, T objectToApplyTo, Operation<T> oper
                         else
                         {
                             throw new JsonPatchException<T>(operationToReport,
-                                string.Format("Patch failed: provided path is invalid for array property type at " +
-                                    "location path: {0}: position larger than array size",
-                                    path, 422),
-                                objectToApplyTo);
+                                Resources.InvalidIndexForArrayProperty(operationToReport.op, path),
```
Comment (pranavkm) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: Did the resx generation change recently? This should be `Resources.FormatInvalidIndexForArrayProperty`

**Comment by kirthik** [comment]
I manually created it but talked to @pranavkm. I will generate it with the command provided.

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 2]
```diff
@@ -158,20 +158,18 @@ private void Add(string path, object value, T objectToApplyTo, Operation<T> oper
...
+                                objectToApplyTo,
+                                422);
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: We should get rid of this. We're never going to have an Exception -> StatusCode feature built-in, and this gives the impression that it does something.

File: test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs [hunk 3]
```diff
@@ -281,7 +281,10 @@ public void AddToListInvalidPositionTooLarge()
             patchDoc.Add<int>(o => o.SimpleDTO.IntegerList, 4, 4);
 
             // Act & Assert
-            Assert.Throws<JsonPatchException<SimpleDTOWithNestedDTO>>(() => { patchDoc.ApplyTo(doc); });
+            var exception = Assert.Throws<JsonPatchException<SimpleDTOWithNestedDTO>>(() => { patchDoc.ApplyTo(doc); });
+            Assert.Equal("For Operation 'add', array property at location path: '/simpledto/integerlist/4' index is " +
+                "larger than array size.", exception.Message);
```
Comment (rynowak) [inline test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs]: let's try to simplify this message and boil it down to something more readable.

Suggestion:
```
"For Operation 'add' on array property '/simpledto/integerlist/4', the index is larger than the array size."
```

File: src/Microsoft.AspNet.JsonPatch/Resources.resx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="CannotUpdateProperty" xml:space="preserve">
+    <value>Cannot update property at path '{0}'.</value>
+  </data>
+  <data name="InvalidIndexForArrayProperty" xml:space="preserve">
+    <value>For Operation '{0}', array property at location path: '{1}' index is larger than array size.</value>
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Resources.resx]: In general with these the single-quotes are enough, you don't need the colon.

Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Resources.resx]: Operation can be lower case.

File: src/Microsoft.AspNet.JsonPatch/Resources.resx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  </resheader>
+  <resheader name="reader">
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="CannotUpdateProperty" xml:space="preserve">
+    <value>Cannot update property at path '{0}'.</value>
+  </data>
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Resources.resx]: What's this one used for? read-only properties? jsonignore?

**Comment by rynowak** [comment]
The change in general looks right, let's cleanup the grammar a bit. :watch:

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs (hunk 2 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: will remove.

File: test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs (hunk 3 above)
Comment (kirthik) [inline test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs]: Ok. I will update the message.

File: src/Microsoft.AspNet.JsonPatch/Resources.resx (hunk 5 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Resources.resx]: for jsonignore

File: src/Microsoft.AspNet.JsonPatch/Resources.resx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  <data name="InvalidIndexForArrayProperty" xml:space="preserve">
+    <value>For operation '{0}' on array property at path '{1}', the index is larger than the array size.</value>
+  </data>
+  <data name="InvalidJsonPatchDocument" xml:space="preserve">
+    <value>The JsonPatchDocument was malformed and could not be parsed.</value>
+  </data>
+  <data name="InvalidPathForArrayProperty" xml:space="preserve">
+    <value>For operation '{0}',  the provided path is invalid for array property at path '{1}'.</value>
```
Comment (Eilon) [inline src/Microsoft.AspNet.JsonPatch/Resources.resx]: One space after the comma

Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Resources.resx]: I will fix it.

**Comment by kirthik** [comment]
@rynowak - Could you please help look at updated PR?

**Comment by kirthik** [comment]
I messed up something during rebase. I am fixing it. Please don't look at this PR until then.

**Comment by kirthik** [comment]
@rynowak - I updated the PR. Could you please take a look?

**Comment by rynowak** [comment]
:shipit: after fixing spacing. 

Is there an easy to test the "The document was invalid and ..." message? I didn't notice a test for it. Can you add something if it's reasonably possible to add in a test.

**Comment by kirthik** [comment]
We have a test `DeserializationMustFailWithEnvelope` in `ObjectAdapterTests.cs`. I added an `Assert` to check the exception message.

**Pull request closed without merging**